src/ucp/tag/offload_proto.c

```diff
@@ -0,0 +1,209 @@
+        !UCP_MEM_IS_HOST(select_param->mem_type)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_single_init(&params);
+}
+
+static ucp_proto_t ucp_eager_tag_offload_short_proto = {
+    .name       = "egr/single/offload/short",
+    .flags      = UCP_PROTO_FLAG_TAG_SHORT,
+    .init       = ucp_proto_eager_tag_offload_short_init,
+    .config_str = ucp_proto_single_config_str,
+    .progress   = ucp_proto_eager_tag_offload_short_progress
+};
+UCP_PROTO_REGISTER(&ucp_eager_tag_offload_short_proto);
+
+static size_t ucp_eager_tag_offload_pack(void *dest, void *arg)
+{
+    ucp_request_t *req = arg;
+    ucp_datatype_iter_t next_iter;
+
+    ucs_assert(req->send.state.dt_iter.offset == 0);
+
+    return ucp_datatype_iter_next_pack(&req->send.state.dt_iter,
+                                       req->send.ep->worker, SIZE_MAX,
+                                       &next_iter, dest);
+}
+
+static ucs_status_t
+ucp_proto_eager_tag_offload_bcopy_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ssize_t packed_len;
+    ucs_status_t status;
+
+    packed_len = uct_ep_tag_eager_bcopy(req->send.ep->uct_eps[spriv->super.lane],
+                                        req->send.msg_proto.tag.tag, 0ul,
+                                        ucp_eager_tag_offload_pack, req, 0);
+    status     = ucs_likely(packed_len >= 0) ? UCS_OK : packed_len;
+
+    return ucp_proto_single_status_handle(req, ucp_proto_request_bcopy_complete,
+                                          ucp_proto_request_bcopy_complete,
+                                          spriv->super.lane, status);
+}
+
+static ucs_status_t ucp_proto_eager_tag_offload_bcopy_init(
+        const ucp_proto_init_params_t *init_params)
+{
+    ucp_context_t *context                = init_params->worker->context;
+    ucp_proto_single_init_params_t params = {
+        .super.super         = *init_params,
+        .super.latency       = 0,
+        .super.overhead      = 5e-9,
+        .super.cfg_thresh    = context->config.ext.bcopy_thresh,
+        .super.cfg_priority  = 20,
+        .super.min_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.max_frag_offs = ucs_offsetof(uct_iface_attr_t,
+                                            cap.tag.eager.max_bcopy),
+        .super.hdr_size      = sizeof(ucp_tag_t),
+        .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_MAX_FRAG |
+                               UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY,
+        .lane_type           = UCP_LANE_TYPE_TAG,
+        .tl_cap_flags        = UCT_IFACE_FLAG_TAG_EAGER_BCOPY
+    };
+
+    if ((init_params->select_param->op_id != UCP_OP_ID_TAG_SEND) ||
+        /* offload proto can not be used if no tag offload lane configured */
+        !ucp_ep_config_key_has_tag_lane(init_params->ep_config_key)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
```

Comment:
is it possible to create some common macro for this check?

Reply:
created